be/src/exec/pipeline/aggregate/aggregate_streaming_sink_operator.cpp

```diff
@@ -89,63 +89,190 @@ Status AggregateStreamingSinkOperator::_push_chunk_by_force_preaggregation(Chunk
-                ChunkPtr res = std::make_shared<Chunk>();
-                RETURN_IF_ERROR(_aggregator->output_chunk_by_streaming_with_selection(chunk.get(), &res));
-                _aggregator->offer_chunk_to_buffer(res);
+
+        } else if (_auto_context.adjust_count < continuous_limit &&
+                   _auto_context.is_high_reduction(agg_count, chunk_size) &&
+                   allocated_bytes < AggrAutoContext::MaxHtSize) {
+            RETURN_IF_ERROR(_push_chunk_by_force_preaggregation(chunk, chunk_size));
+
+            _auto_context.preagg_count++;
+            _auto_context.pass_through_count = 0;
+            _auto_context.selective_preagg_count = 0;
+            if (_auto_context.preagg_count == AggrAutoContext::StableLimit) {
+                _auto_state = AggrAutoState::PREAGG;
+                _auto_context.preagg_count = 0;
+                VLOG_ROW << "auto agg: continuous " << AggrAutoContext::StableLimit << " high reduction "
+                         << agg_count * 1.0 / chunk_size << " "
+                         << _auto_context.get_auto_state_string(AggrAutoState::ADJUST) << " -> "
+                         << _auto_context.get_auto_state_string(_auto_state);
+            }
+        } else {
+            RETURN_IF_ERROR(_push_chunk_by_selective_preaggregation(chunk, chunk_size, false));
+            _auto_context.selective_preagg_count++;
+            _auto_context.pass_through_count = 0;
+            _auto_context.preagg_count = 0;
+            if (_auto_context.selective_preagg_count == AggrAutoContext::StableLimit) {
+                _auto_state = AggrAutoState::SELECTIVE_PREAGG;
+                VLOG_ROW << "auto agg: continuous " << AggrAutoContext::StableLimit << " "
+                         << _auto_context.get_auto_state_string(AggrAutoState::ADJUST)
+                         << _auto_context.get_auto_state_string(_auto_state);
             }
         }
-
-        COUNTER_SET(_aggregator->hash_table_size(), (int64_t)_aggregator->hash_map_variant().size());
+        break;
     }
+    case AggrAutoState::PASS_THROUGH: {
+        RETURN_IF_ERROR(_push_chunk_by_force_streaming(chunk));
+        _auto_context.pass_through_count++;
+        if (_auto_context.pass_through_count > continuous_limit) {
+            _auto_state =
+                    allocated_bytes < AggrAutoContext::MaxHtSize ? AggrAutoState::FORCE_PREAGG : AggrAutoState::ADJUST;
+            _auto_context.pass_through_count = 0;
+            _auto_context.preagg_count = 0;
+            _auto_context.adjust_count = 0;
 
+            VLOG_ROW << "auto agg: continuous " << continuous_limit << " "
+                     << _auto_context.get_auto_state_string(AggrAutoState::PASS_THROUGH) << " -> "
+                     << _auto_context.get_auto_state_string(_auto_state);
+            _auto_context.update_continuous_limit();
+        }
+        break;
+    }
+    case AggrAutoState::FORCE_PREAGG:
```

Comment:
can FORCE_PREAGG be a converged stable stage if preagg have a very significant reduction? for examples, group-by columns are low-cardinality column, such as:
1.  gender of student, 
2. lo_shipmode, lo_shippriority of ssb.lineorder.

Reply:
of cause, the reduction ratio >= HighReduction, it will be FORCE_PREAGG, also verified in TPCDS/TPCH.